fem/hybridization.cpp

```diff
@@ -128,6 +130,69 @@ void Hybridization::ConstructC()
          elmat.Threshold(1e-12 * elmat.MaxMaxNorm());
          Ct->AddSubMatrix(vdofs, c_vdofs, elmat, skip_zeros);
       }
+
+      if (c_bfbfi.Size())
+      {
+         const FiniteElement *fe1, *fe2;
+         const FiniteElement *face_el;
+
+         // Which boundary attributes need to be processed?
+         Array<int> bdr_attr_marker(mesh->bdr_attributes.Size() ?
+                                    mesh->bdr_attributes.Max() : 0);
+         bdr_attr_marker = 0;
+         for (int k = 0; k < c_bfbfi.Size(); k++)
+         {
+            if (c_bfbfi_marker[k] == NULL)
+            {
+               bdr_attr_marker = 1;
+               break;
+            }
+            Array<int> &bdr_marker = *c_bfbfi_marker[k];
+            MFEM_ASSERT(bdr_marker.Size() == bdr_attr_marker.Size(),
+                        "invalid boundary marker for boundary face integrator #"
+                        << k << ", counting from zero");
+            for (int i = 0; i < bdr_attr_marker.Size(); i++)
+            {
+               bdr_attr_marker[i] |= bdr_marker[i];
+            }
+         }
+
+         for (int i = 0; i < fes->GetNBE(); i++)
+         {
+            const int bdr_attr = mesh->GetBdrAttribute(i);
+            if (bdr_attr_marker[bdr_attr-1] == 0) { continue; }
+
+            FTr = mesh->GetBdrFaceTransformations(i);
+            if (!FTr) { continue; }
+
+            int o1 = hat_offsets[FTr->Elem1No];
+            int s1 = hat_offsets[FTr->Elem1No+1] - o1;
+
+            vdofs.SetSize(s1);
+            for (int j = 0; j < s1; j++)
+            {
+               vdofs[j] = o1 + j;
+            }
+            int iface = mesh->GetBdrElementFaceIndex(i);
+            c_fes->GetFaceVDofs(iface, c_vdofs);
+            face_el = c_fes->GetFaceElement(iface);
+            fe1 = fes -> GetFE (FTr -> Elem1No);
+            // The fe2 object is really a dummy and not used on the boundaries,
+            // but we can't dereference a NULL pointer, and we don't want to
+            // actually make a fake element.
+            fe2 = fe1;
+            for (int k = 0; k < c_bfbfi.Size(); k++)
+            {
+               if (c_bfbfi_marker[k] &&
+                   (*c_bfbfi_marker[k])[bdr_attr-1] == 0) { continue; }
+
+               c_bfbfi[k]->AssembleFaceMatrix(*face_el, *fe1, *fe2, *FTr, elmat);
+               // zero-out small elements in elmat
+               elmat.Threshold(1e-12 * elmat.MaxMaxNorm());
```

Comment:
The static `1e-12` for `real_t == double` works, but you might consider something more dynamic if you plan to support `real_t == float`

Reply:
This is a good point 👍 . I copied that from above (integration of the interior constraint). After a quick search, it turned out this particular number `1e-12` appears at multiple places in the library. I do not know, it would be rather a subject for a separate PR to handle this globally. Or maybe not, because it might depend on the context, so it is better to keep it local even if the numbers will be the same in the end 🤔 . In any case, what should be the number? Taking it relatively that 1e-12 is 3/4 in exponent from double eps 1e-16, it gives me for single eps 6e-8 number approximately 4e-6. So would you take 4e-6?

Reply:
Ok, I added that 4e-6 locally for now :wink:.

Reply:
I personally don't have much experience with running in single precision, but your motivation for `4e-6` seems reasonable to me.  I suppose you could also have the code throw an error for now if `MFEM_USE_SINGLE` is true and decide on the value when there's more support for testing it properly.  Either approach sounds good to me so I'll resolve this thread.